ghcide/session-loader/Development/IDE/Session.hs

```diff
@@ -91,6 +105,58 @@ defaultLoadingOptions = SessionLoadingOptions
     ,getCacheDirs = getCacheDirsDefault
     }
 
+-- | Sets `unsafeGlobalDynFlags` on using the hie-bios cradle and returns the GHC libdir
+setInitialDynFlags :: IO (Maybe LibDir)
+setInitialDynFlags = do
+  dir <- IO.getCurrentDirectory
+  hieYaml <- runMaybeT $ yamlConfig dir
+  cradle <- maybe (HieBios.loadImplicitCradle $ addTrailingPathSeparator dir) HieBios.loadCradle hieYaml
+  libDirRes <- getRuntimeGhcLibDir cradle
+  libdir <- case libDirRes of
+      CradleSuccess libdir -> pure $ Just $ LibDir libdir
+      CradleFail err -> do
+        hPutStrLn stderr $ "Couldn't load cradle for libdir: " ++ show (err,dir,hieYaml,cradle)
+        pure Nothing
+      CradleNone -> do
+        hPutStrLn stderr $ "Couldn't load cradle (CradleNone)"
+        pure Nothing
+  dynFlags <- mapM dynFlagsForPrinting libdir
+  mapM_ setUnsafeGlobalDynFlags dynFlags
+  pure libdir
+
+-- | Wraps `withHieDb` to provide a database connection for reading, and a `HieWriterChan` for
+-- writing. Actions are picked off one by one from the `HieWriterChan` and executed in serial
+-- by a worker thread using a dedicated database connection.
+-- This is done in order to serialize writes to the database, or else SQLite becomes unhappy
+runWithDb :: FilePath -> (HieDb -> IndexQueue -> IO ()) -> IO ()
+runWithDb fp k = do
+  -- Delete the database if it has an incompatible schema version
+  withHieDb fp (const $ pure ())
+    `catch` \IncompatibleSchemaVersion{} -> removeFile fp
+  withHieDb fp $ \writedb -> do
+    initConn writedb
+    _ <- garbageCollectTypeNames writedb
```

Comment:
how expensive is `garbageCollectTypeNames`? Could it be moved to the worker thread?

Reply:
Not very expensive at all. It should be instant for most databases we encounter. It used take some time (~1s) for ~500 .hie files, but I added an additional index which sped it up considerably (https://github.com/wz1000/HieDb/blob/master/src/HieDb/Create.hs#L145).

I think it is fine here, since it will be called once per startup. The savings due to this garbage collection aren't too significant (a few kb), but the garbage can build up as files are indexed and re-indexed.

It is necessary since whenever we encounter a type constructor in a file, we add it to a global symbol table (all other tables in the database are keyed by file, so garbage building up for them isn't an issue since we delete all information associated with a file when re-indexing). The type references table then references the constructors in this symbol table for each type reference. `garbageCollectTypeNames` just deletes all entries in the symbol table that have no incoming references. 

The entries can build up as types are deleted/renamed/moved (there will still be an entry in the symbol table), types are no longer referenced and so on.

Since it still could have a non-trivial cost since it traverses all rows in the table we don't want to call it after each indexing operation. We could call it when indexing is done (i.e. when we send the "progress done" notification for indexing in `indexHieFile`), but since garbage builds up so slowly I think it is good enough to just call it once per startup.

Reply:
Also, I couldn't not measure any additional cost while indexing due to the extra index table to make this operation faster.

Reply:
Sorry I wasn't clear, my suggestion was only to make the call async, not to call it more often

Reply:
I really don't think it would make much of a noticeable difference.

Reply:
How long would it take if there were, say, 30.000 `.hie` files?

Reply:
Only one way to find out: https://github.com/wz1000/HieDb/commit/82e3e96adc241fd93a6ee184e9cd25b1b6bea827

`hiedb index dir-with-hie-files/`

Reply:
How difficult would it be to make this call asynchronous?

Reply:
Done.